File: src/falcon_core/physics/units/units.py
"""Convenient access to common units with symbolic representation."""

from .common_units import CommonUnits
from .symbol_unit import SymbolUnit


class Units:
    """Provides convenient access to common units as SymbolUnit objects.

    This class provides static attributes for all common units, allowing
    easy access like Units.VOLT instead of SymbolUnit(CommonUnits.VOLT).

    All units are pre-defined as class attributes for direct access with
    IDE completion support.
    """

    # SI base units
    METER = SymbolUnit(CommonUnits.METER)
    KILOGRAM = SymbolUnit(CommonUnits.KILOGRAM)
    SECOND = SymbolUnit(CommonUnits.SECOND)
    AMPERE = SymbolUnit(CommonUnits.AMPERE)
    KELVIN = SymbolUnit(CommonUnits.KELVIN)
    MOLE = SymbolUnit(CommonUnits.MOLE)
    CANDELA = SymbolUnit(CommonUnits.CANDELA)

    # Derived SI units
    HERTZ = SymbolUnit(CommonUnits.HERTZ)
    NEWTON = SymbolUnit(CommonUnits.NEWTON)
    PASCAL = SymbolUnit(CommonUnits.PASCAL)
    JOULE = SymbolUnit(CommonUnits.JOULE)
    WATT = SymbolUnit(CommonUnits.WATT)
    COULOMB = SymbolUnit(CommonUnits.COULOMB)
    VOLT = SymbolUnit(CommonUnits.VOLT)
    FARAD = SymbolUnit(CommonUnits.FARAD)
    OHM = SymbolUnit(CommonUnits.OHM)
    SIEMENS = SymbolUnit(CommonUnits.SIEMENS)
    WEBER = SymbolUnit(CommonUnits.WEBER)
    TESLA = SymbolUnit(CommonUnits.TESLA)
    HENRY = SymbolUnit(CommonUnits.HENRY)

    # Common non-SI units
    MINUTE = SymbolUnit(CommonUnits.MINUTE)
    HOUR = SymbolUnit(CommonUnits.HOUR)
    ELECTRON_VOLT = SymbolUnit(CommonUnits.ELECTRON_VOLT)

    # Temperature units with offsets
    CELSIUS = SymbolUnit(CommonUnits.CELSIUS)
    FAHRENHEIT = SymbolUnit(CommonUnits.FAHRENHEIT)

    # Dimensionless units
    DIMENSIONLESS = SymbolUnit(CommonUnits.DIMENSIONLESS)
    PERCENT = SymbolUnit(CommonUnits.PERCENT)
    RADIAN = SymbolUnit(CommonUnits.RADIAN)

    # Common prefixed units
    MILLIMETER = SymbolUnit(CommonUnits.get_milli(CommonUnits.METER))
    CENTIMETER = SymbolUnit(CommonUnits.get_unit_with_prefix(CommonUnits.METER, "c"))
    KILOMETER = SymbolUnit(CommonUnits.get_kilo(CommonUnits.METER))

    MILLIVOLT = SymbolUnit(CommonUnits.get_milli(CommonUnits.VOLT))
    KILOVOLT = SymbolUnit(CommonUnits.get_kilo(CommonUnits.VOLT))

    MILLIAMPERE = SymbolUnit(CommonUnits.get_milli(CommonUnits.AMPERE))
    MICROAMPERE = SymbolUnit(CommonUnits.get_micro(CommonUnits.AMPERE))

    MILLISECOND = SymbolUnit(CommonUnits.get_milli(CommonUnits.SECOND))
    MICROSECOND = SymbolUnit(CommonUnits.get_micro(CommonUnits.SECOND))

    MILLIOHM = SymbolUnit(CommonUnits.get_milli(CommonUnits.OHM))
    KILOOHM = SymbolUnit(CommonUnits.get_kilo(CommonUnits.OHM))
    MEGAOHM = SymbolUnit(CommonUnits.get_mega(CommonUnits.OHM))

    MILLIHERTZ = SymbolUnit(CommonUnits.get_milli(CommonUnits.HERTZ))
    KILOHERTZ = SymbolUnit(CommonUnits.get_kilo(CommonUnits.HERTZ))
    MEGAHERTZ = SymbolUnit(CommonUnits.get_mega(CommonUnits.HERTZ))

    # Derived units from combinations
    METERS_PER_SECOND = SymbolUnit(CommonUnits.METER / CommonUnits.SECOND)
    METERS_PER_SECOND_SQUARED = SymbolUnit(CommonUnits.METER / (CommonUnits.SECOND**2))

    NEWTON_METER = SymbolUnit(CommonUnits.NEWTON * CommonUnits.METER)
    NEWTON_PER_METER = SymbolUnit(CommonUnits.NEWTON / CommonUnits.METER)

    VOLT_PER_METER = SymbolUnit(CommonUnits.VOLT / CommonUnits.METER)

    AMPERE_PER_VOLT = SymbolUnit(CommonUnits.AMPERE / CommonUnits.VOLT)  # Siemens
    VOLT_PER_AMPERE = SymbolUnit(CommonUnits.VOLT / CommonUnits.AMPERE)  # Ohm

    WATTS_PER_METER_KELVIN = SymbolUnit(
        CommonUnits.WATT / (CommonUnits.METER * CommonUnits.KELVIN)
    )

    # Convenience method to get a unit by name
    @classmethod
    def get(cls, unit_name: str) -> SymbolUnit:
        """Get a unit by name.

        Args:
            unit_name: The name of the unit (case-insensitive)

        Returns:
            The corresponding SymbolUnit

        Raises:
            ValueError: If the unit is not found
        """
        try:
            return getattr(cls, unit_name.upper())
        except AttributeError:
            msg = f"Unit '{unit_name}' not found"
            raise ValueError(msg)

    @classmethod
    def list_available(cls) -> list[str]:
        """List all available unit names.

        Returns:
            A list of available unit names
        """
        return [
            attr
            for attr in dir(cls)
            if not attr.startswith("_") and isinstance(getattr(cls, attr), SymbolUnit)
        ]

File: src/physics/units/Units.cpp
#include "falcon_core/physics/units/Units.hpp"

#include "falcon_core/physics/units/CommonUnits.hpp"
#include "falcon_core/physics/units/SymbolUnit.hpp"
#include "falcon_core/physics/units/Unit.hpp"

using namespace falcon_core::physics::units;
const SymbolUnit Units::Meter{CommonUnits::Meter};
const SymbolUnit Units::Kilogram{CommonUnits::Kilogram};
const SymbolUnit Units::Second{CommonUnits::Second};
const SymbolUnit Units::Ampere{CommonUnits::Ampere};
const SymbolUnit Units::Kelvin{CommonUnits::Kelvin};
const SymbolUnit Units::Mole{CommonUnits::Mole};
const SymbolUnit Units::Candela{CommonUnits::Candela};
const SymbolUnit Units::Hertz{CommonUnits::Hertz};
const SymbolUnit Units::Newton{CommonUnits::Newton};
const SymbolUnit Units::Pascal{CommonUnits::Pascal};
const SymbolUnit Units::Joule{CommonUnits::Joule};
const SymbolUnit Units::Watt{CommonUnits::Watt};
const SymbolUnit Units::Coulomb{CommonUnits::Coulomb};
const SymbolUnit Units::Volt{CommonUnits::Volt};
const SymbolUnit Units::Farad{CommonUnits::Farad};
const SymbolUnit Units::Ohm{CommonUnits::Ohm};
const SymbolUnit Units::Siemens{CommonUnits::Siemens};
const SymbolUnit Units::Weber{CommonUnits::Weber};
const SymbolUnit Units::Tesla{CommonUnits::Tesla};
const SymbolUnit Units::Henry{CommonUnits::Henry};
const SymbolUnit Units::Minute{CommonUnits::Minute};
const SymbolUnit Units::Hour{CommonUnits::Hour};
const SymbolUnit Units::ElectronVolt{CommonUnits::ElectronVolt};
const SymbolUnit Units::Celsius{CommonUnits::Celsius};
const SymbolUnit Units::Fahrenheit{CommonUnits::Fahrenheit};
const SymbolUnit Units::Dimensionless{CommonUnits::Dimensionless};
const SymbolUnit Units::Percent{CommonUnits::Percent};
const SymbolUnit Units::Radian{CommonUnits::Radian};
const SymbolUnit Units::MilliMeter{CommonUnits::get_milli(CommonUnits::Meter)};
const SymbolUnit Units::KiloMeter{CommonUnits::get_kilo(CommonUnits::Meter)};
const SymbolUnit Units::MilliVolt{CommonUnits::get_milli(CommonUnits::Volt)};
const SymbolUnit Units::KiloVolt{CommonUnits::get_kilo(CommonUnits::Volt)};
const SymbolUnit Units::MilliAmpere{
    CommonUnits::get_milli(CommonUnits::Ampere)};
const SymbolUnit Units::MicroAmpere{
    CommonUnits::get_micro(CommonUnits::Ampere)};
const SymbolUnit Units::NanoAmpere{CommonUnits::get_nano(CommonUnits::Ampere)};
const SymbolUnit Units::PicoAmpere{CommonUnits::get_pico(CommonUnits::Ampere)};
const SymbolUnit Units::MilliSecond{
    CommonUnits::get_milli(CommonUnits::Second)};
const SymbolUnit Units::MicroSecond{
    CommonUnits::get_micro(CommonUnits::Second)};
const SymbolUnit Units::NanoSecond{CommonUnits::get_nano(CommonUnits::Second)};
const SymbolUnit Units::PicoSecond{CommonUnits::get_pico(CommonUnits::Second)};
const SymbolUnit Units::MilliOhm{CommonUnits::get_milli(CommonUnits::Ohm)};
const SymbolUnit Units::KiloOhm{CommonUnits::get_kilo(CommonUnits::Ohm)};
const SymbolUnit Units::MegaOhm{CommonUnits::get_mega(CommonUnits::Ohm)};
const SymbolUnit Units::MilliHertz{CommonUnits::get_milli(CommonUnits::Hertz)};
const SymbolUnit Units::KiloHertz{CommonUnits::get_kilo(CommonUnits::Hertz)};
const SymbolUnit Units::MegaHertz{CommonUnits::get_mega(CommonUnits::Hertz)};
const SymbolUnit Units::GigaHertz{CommonUnits::get_giga(CommonUnits::Hertz)};
const SymbolUnit Units::MetersPerSecond{*CommonUnits::Meter /
                                        CommonUnits::Second};
const SymbolUnit Units::MetersPerSecondSquared{*CommonUnits::Meter /
                                               (*CommonUnits::Second ^ 2)};
const SymbolUnit Units::NewtonMeter{*CommonUnits::Newton * CommonUnits::Meter};
const SymbolUnit Units::NewtonPerMeter{*CommonUnits::Newton /
                                       CommonUnits::Meter};
const SymbolUnit Units::VoltsPerMeter{*CommonUnits::Volt / CommonUnits::Meter};
const SymbolUnit Units::VoltsPerSecond{*CommonUnits::Volt /
                                       CommonUnits::Second};
const SymbolUnit Units::AmperesPerMeter{*CommonUnits::Ampere /
                                        CommonUnits::Meter};
const SymbolUnit Units::VoltsPerAmpere{*CommonUnits::Volt /
                                       CommonUnits::Ampere};
const SymbolUnit Units::WattsPerMeterKelvin{
    *CommonUnits::Watt / (*CommonUnits::Meter * CommonUnits::Kelvin)};
